DefaultTCPIPHandler executes before the agent and server connect to each other

### Description

## Background

I am using a database migration tool called Airbyte. I intend to access the database in the private network as shown in the diagram below by forming an SSH tunnel through ShellHub and via ShellHub agent.

![image](https://www.plantuml.com/plantuml/svg/POrHQiCm44J_TOeX_yu9GibLrjRM4f5tnMgTKaBkNbMfeP0F3HfstiorEPdlMmdjbgMIqOQIXVsxCjGyQIZywN25vJKvOZPUFAk4yFiXMGlZL2tVoHd2_gLsE-4HW4XECpL6dBiJZGoEAuyRiEXMLLWSj7ODm3Cy_vSkHVU8nu6snAMaVNx3Mqi3Fd9iTxRtaLgFYJZZCar_iqCnJH-Zv-Kkv_Dm_RsIdGrNbhXlvGS0)

<img width="1677" alt="Screenshot 2023-11-01 at 12 34 02" src="https://github.com/shellhub-io/shellhub/assets/15084045/fb5da8bb-4792-4c1d-8bcd-6bc63fdc9a11">

## Problem

> connection between server and agent is not established yet

```
Failed (SshChannelOpenException) to open channel for session=MinaSession[local=/127.0.0.1:36247, remote=/127.0.0.1:49208]: connection between server and agent is not established yet
```

<details>
<summary>Learn more</summary>

```
Internal message: io.airbyte.commons.exceptions.ConnectionErrorException: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 10002ms.
Failure origin: source
Failure type: config_error
```

</details>

<details>
<summary>Stacktrace</summary>

```
io.airbyte.commons.exceptions.ConnectionErrorException: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 10002ms.
	at io.airbyte.db.jdbc.DefaultJdbcDatabase.getMetaData(DefaultJdbcDatabase.java:87)
	at io.airbyte.integrations.source.postgres.PostgresSource.createDatabase(PostgresSource.java:315)
	at io.airbyte.integrations.source.postgres.PostgresSource.createDatabase(PostgresSource.java:121)
	at io.airbyte.integrations.source.relationaldb.AbstractDbSource.check(AbstractDbSource.java:85)
	at io.airbyte.integrations.source.postgres.PostgresSource.check(PostgresSource.java:702)
	at io.airbyte.integrations.base.ssh.SshTunnel.sshWrap(SshTunnel.java:270)
	at io.airbyte.integrations.base.ssh.SshWrappedSource.check(SshWrappedSource.java:54)
	at io.airbyte.integrations.base.IntegrationRunner.runInternal(IntegrationRunner.java:148)
	at io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:123)
	at io.airbyte.integrations.source.postgres.PostgresSource.main(PostgresSource.java:683)
Caused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 10002ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:181)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:146)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100)
	at io.airbyte.db.jdbc.DefaultJdbcDatabase.getMetaData(DefaultJdbcDatabase.java:78)
	... 9 more
Caused by: org.postgresql.util.PSQLException: Connection to 127.0.0.1:36247 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:319)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:247)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:733)
	at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:712)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at org.postgresql.core.PGStream.createSocket(PGStream.java:241)
	at org.postgresql.core.PGStream.<init>(PGStream.java:98)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:109)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235)
	... 14 more
```

</details>

<details>
<summary>Logs</summary>

```
2023-11-01 03:33:29 i.a.w.t.TemporalAttemptExecution(get):124 - Docker volume job log path: /tmp/workspace/1cf4099a-53ca-454d-91c1-39d15b591d21/0/logs.log
2023-11-01 03:33:29 i.a.w.t.TemporalAttemptExecution(get):129 - Executing worker wrapper. Airbyte version: 0.50.29
2023-11-01 03:33:29 i.a.a.c.AirbyteApiClient(retryWithJitterThrows):290 - Attempt 0 to save workflow id for cancellation
2023-11-01 03:33:29 i.a.c.i.LineGobbler(voidCall):149 - 
2023-11-01 03:33:29 i.a.c.EnvConfigs(getEnvOrDefault):1228 - Using default value for environment variable SIDECAR_KUBE_CPU_LIMIT: '2.0'
2023-11-01 03:33:29 i.a.c.i.LineGobbler(voidCall):149 - ----- START CHECK -----
2023-11-01 03:33:29 i.a.c.EnvConfigs(getEnvOrDefault):1228 - Using default value for environment variable SOCAT_KUBE_CPU_LIMIT: '2.0'
2023-11-01 03:33:29 i.a.c.i.LineGobbler(voidCall):149 - 
2023-11-01 03:33:29 i.a.c.EnvConfigs(getEnvOrDefault):1228 - Using default value for environment variable SIDECAR_KUBE_CPU_REQUEST: '0.1'
2023-11-01 03:33:29 i.a.c.EnvConfigs(getEnvOrDefault):1228 - Using default value for environment variable SOCAT_KUBE_CPU_REQUEST: '0.1'
2023-11-01 03:33:29 i.a.c.EnvConfigs(getEnvOrDefault):1228 - Using default value for environment variable LAUNCHDARKLY_KEY: ''
2023-11-01 03:33:29 i.a.c.i.LineGobbler(voidCall):149 - Checking if airbyte/source-postgres:3.1.7 exists...
2023-11-01 03:33:29 i.a.w.p.DockerProcessFactory(create):143 - Creating docker container = source-postgres-check-1cf4099a-53ca-454d-91c1-39d15b591d21-0-xazys with resources io.airbyte.config.ResourceRequirements@10f23d9c[cpuRequest=,cpuLimit=,memoryRequest=,memoryLimit=,additionalProperties={}] and allowedHosts null
2023-11-01 03:33:29 i.a.c.i.LineGobbler(voidCall):149 - airbyte/source-postgres:3.1.7 was found locally.
2023-11-01 03:33:29 i.a.w.p.DockerProcessFactory(create):196 - Preparing command: docker run --rm --init -i -w /data/1cf4099a-53ca-454d-91c1-39d15b591d21/0 --log-driver none --name source-postgres-check-1cf4099a-53ca-454d-91c1-39d15b591d21-0-xazys --network host -v airbyte_workspace:/data -v /tmp/airbyte_local:/local -e DEPLOYMENT_MODE=OSS -e WORKER_CONNECTOR_IMAGE=airbyte/source-postgres:3.1.7 -e AUTO_DETECT_SCHEMA=true -e LAUNCHDARKLY_KEY= -e SOCAT_KUBE_CPU_REQUEST=0.1 -e SOCAT_KUBE_CPU_LIMIT=2.0 -e USE_STREAM_CAPABLE_STATE=true -e FIELD_SELECTION_WORKSPACES= -e WORKER_ENVIRONMENT=DOCKER -e AIRBYTE_ROLE= -e APPLY_FIELD_SELECTION=false -e WORKER_JOB_ATTEMPT=0 -e OTEL_COLLECTOR_ENDPOINT=http://host.docker.internal:4317 -e FEATURE_FLAG_CLIENT=config -e AIRBYTE_VERSION=0.50.29 -e WORKER_JOB_ID=1cf4099a-53ca-454d-91c1-39d15b591d21 airbyte/source-postgres:3.1.7 check --config source_config.json
2023-11-01 03:33:29 i.a.w.i.VersionedAirbyteStreamFactory(create):177 - Reading messages from protocol version 0.2.0
2023-11-01 03:33:31 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.s.p.PostgresSource(main):682 starting source: class io.airbyte.integrations.source.postgres.PostgresSource
2023-11-01 03:33:31 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.b.IntegrationCliParser(parseOptions):126 integration args: {check=null, config=source_config.json}
2023-11-01 03:33:31 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.b.IntegrationRunner(runInternal):130 Running integration: io.airbyte.integrations.base.ssh.SshWrappedSource
2023-11-01 03:33:31 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.b.IntegrationRunner(runInternal):131 Command: CHECK
2023-11-01 03:33:31 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.b.IntegrationRunner(runInternal):132 Integration config: IntegrationConfig{command=CHECK, configPath='source_config.json', catalogPath='null', statePath='null'}
2023-11-01 03:33:32 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN c.n.s.JsonMetaSchema(newValidator):278 Unknown keyword groups - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2023-11-01 03:33:32 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN c.n.s.JsonMetaSchema(newValidator):278 Unknown keyword order - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2023-11-01 03:33:32 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN c.n.s.JsonMetaSchema(newValidator):278 Unknown keyword group - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2023-11-01 03:33:32 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN c.n.s.JsonMetaSchema(newValidator):278 Unknown keyword airbyte_secret - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2023-11-01 03:33:32 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN c.n.s.JsonMetaSchema(newValidator):278 Unknown keyword always_show - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2023-11-01 03:33:32 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN c.n.s.JsonMetaSchema(newValidator):278 Unknown keyword display_type - you should define your own Meta Schema. If the keyword is irrelevant for validation, just use a NonValidationKeyword
2023-11-01 03:33:32 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.b.s.SshTunnel(getInstance):204 Starting connection with method: SSH_PASSWORD_AUTH
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO o.a.s.c.u.s.AbstractSecurityProviderRegistrar(getOrCreateProvider):112 getOrCreateProvider(EdDSA) created instance of net.i2p.crypto.eddsa.EdDSASecurityProvider
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO o.a.s.c.i.DefaultIoServiceFactoryFactory(getIoServiceProvider):71 Using MinaServiceFactoryFactory
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN o.a.s.c.k.StaticServerKeyVerifier(handleAcceptance):59 Server at shellhub.xxx.xxx/x.x.x.x:22 presented unverified RSA key: SHA256:vhHKg9OLlLbUICa0HRdop7hCG6VoKgG6Xtmr9BDK1UQ
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.b.s.SshTunnel(openTunnel):367 Established tunneling session to y.y.y.y:5432. Port forwarding started on /127.0.0.1:36247 
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.s.p.PostgresUtils(isCdc):61 using CDC: false
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.s.p.PostgresSource(toSslJdbcParamInternal):719 DISABLED toSslJdbcParam disable
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO c.z.h.HikariDataSource(<init>):80 HikariPool-1 - Starting...
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO c.z.h.HikariDataSource(<init>):82 HikariPool-1 - Start completed.
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN o.a.s.c.c.ClientChannelPendingMessagesQueue(operationComplete):197 operationComplete(ClientChannelPendingMessagesQueue[channel=TcpipClientChannel[id=0, recipient=-1]-ClientSessionImpl[ubuntu@SSHID@shellhub.(redacted domain)/x.x.x.x:22], open=false]) SshChannelOpenException[connection between server and agent is not established yet] signaled
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN o.a.s.c.u.l.LoggingUtils(warn):610 Failed (SshChannelOpenException) to open channel for session=MinaSession[local=/127.0.0.1:36247, remote=/127.0.0.1:49208]: connection between server and agent is not established yet
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN o.a.s.c.c.ClientChannelPendingMessagesQueue(operationComplete):197 operationComplete(ClientChannelPendingMessagesQueue[channel=TcpipClientChannel[id=1, recipient=-1]-ClientSessionImpl[ubuntu@SSHID@shellhub.(redacted domain)/x.x.x.x:22], open=false]) SshChannelOpenException[connection between server and agent is not established yet] signaled
2023-11-01 03:33:33 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN o.a.s.c.u.l.LoggingUtils(warn):610 Failed (SshChannelOpenException) to open channel for session=MinaSession[local=/127.0.0.1:36247, remote=/127.0.0.1:49214]: connection between server and agent is not established yet
2023-11-01 03:33:34 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN o.a.s.c.c.ClientChannelPendingMessagesQueue(operationComplete):197 operationComplete(ClientChannelPendingMessagesQueue[channel=TcpipClientChannel[id=2, recipient=-1]-ClientSessionImpl[ubuntu@SSHID@shellhub.(redacted domain)/x.x.x.x:22], open=false]) SshChannelOpenException[connection between server and agent is not established yet] signaled
2023-11-01 03:33:34 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):309 - WARN o.a.s.c.u.l.LoggingUtils(warn):610 Failed (SshChannelOpenException) to open channel for session=MinaSession[local=/127.0.0.1:36247, remote=/127.0.0.1:49224]: connection between server and agent is not established yet
2023-11-01 03:33:43 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO c.z.h.HikariDataSource(close):350 HikariPool-1 - Shutdown initiated...
2023-11-01 03:33:44 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO c.z.h.HikariDataSource(close):352 HikariPool-1 - Shutdown completed.
2023-11-01 03:33:44 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.b.IntegrationRunner(runInternal):226 Completed integration: io.airbyte.integrations.base.ssh.SshWrappedSource
2023-11-01 03:33:44 i.a.w.i.VersionedAirbyteStreamFactory(internalLog):312 - INFO i.a.i.s.p.PostgresSource(main):684 completed source: class io.airbyte.integrations.source.postgres.PostgresSource
2023-11-01 03:33:44 i.a.w.g.DefaultCheckConnectionWorker(run):117 - Check connection job received output: io.airbyte.config.StandardCheckConnectionOutput@75c877ae[status=failed,message=State code: 08001; Message: Connection to 127.0.0.1:36247 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.,additionalProperties={}]
2023-11-01 03:33:44 i.a.c.i.LineGobbler(voidCall):149 - 
2023-11-01 03:33:44 i.a.c.i.LineGobbler(voidCall):149 - ----- END CHECK -----
2023-11-01 03:33:44 i.a.c.i.LineGobbler(voidCall):149 - 
```
</details>

<img width="646" alt="Screenshot 2023-11-01 at 12 53 00" src="https://github.com/shellhub-io/shellhub/assets/15084045/ac3a73d6-6644-4657-a7e1-c1c529df519e">

## What I have found

I traced the source code to track the SSH tunneling connection process of the Shellhub SSH server and followed the path leading up to DefaultTCPIPHandler.

![image](https://www.plantuml.com/plantuml/svg/ZLPFRzis5B_hKxnTMcTrf6Adme261bxN12YweCbb28A1bbY924rg9BKu7GRiljE03Hjsxd677lUPkkmxZ4zyfcbRNgQ3HRx7z-Fl_NdFUcuTCwvVokoJph1coS1n_GwSwNcMBVGA6k7QVf5NUbdObalfDyT2ntK6WAUiRVqAeFDWeX8TKmwMJ2Wl30UwcmRNj7EzKboIODlgCbLnLUFvn7B97QkPOwEsXbkd3TytJjR4bvllkCds8IwYWavMBKD2TlH2LtMYEyZLMzQIcr5B2jY3DAqLMetn94rgbQxIcEEQTI9X9fJZHZ5PIS6L6oKu9zM375ldEihdeMm2it3gZuNT0SxqtYswHegQ6yaMjWWnsyyMmdjTxPMFns9q0BaJXjkYqSUTqQix_-Cq1IuHx3hqIbZ75JU7Y6l9L9DhqnIheZ9tdTD4EsZ3Uio9kLPg13GhIpXJmWacnJiEPPbbbU7CSQ1WOfV1yUUqXITFPjHTC2lXDRyDyicHnoR3cDcWyiOaWhLfVd9-1kLEbUGUB9ppO14jpMk2c8WtgBGlavB9KuQ6kzweRDyb3s5qxycWJsJuahitWymR_h-BSpg_kKXfnq5Z1KKFJbv_3GMrdIrclQYBIchBeygs1TBt-vSlBg6uNPVb-LSNcoqfMywaQwkMLpT3Gl6-B9DQTt3LQ7H9A7xjJvF7uJqRufE7IZjHzL0Vazcpfy-UJYaEHmWNAu5c5e7GDeGjp3agXQ2OR0K5WqS6tjHp3tQdEBg28Ksn_EBYbAGJ5qq16i-eWxUzgYQXJY1qh8VAgG6Fu6eZ8unh6XRHXNJG0H91FYH6H_mAsL1gi6v8YVcd9qHs8NLr8rH3LwHu-9f1fJtzodc79fLRJI7mdWxOKyZpV8C8k8qx3Ff_Cio_54hODbAgkNL6t-rEwn0qlwD2ZDr4Osmx8tNCOpxiIdWKWekYxsEgVeXbVRWGTejgE_kqce4SJeNqA-x8QZBxp3_984bcOGhpAbg7GOeeZt3bXf5AQ4YTxa7kXzAt5ulK6nJamPFXfatU7iFzRtz-_El3nmy_mltxd-__-EMVNt__-wVtUPg_EW5TbpoEuGdwZ0l2Imir7TRq96Y9ceZQ5ZVJuJF25vVXt_fVRZ2quUyWL39PfSKSgsF_5YgATjEaGoJcuLE52i9ULUrQT7dN-MHVxqE5BnYFrUWrN48XxMxuWhzblNINy_Epy_KWeF7oQThoA9d2DrweDu8G1CLlaT36SQBdm-DMoTpmfnrVkNHkx3YNkWBaXFU6FW1X5hyJypT1zC8wjf32jhoUx5yKqTBs74r3EX18GLPRNJ3I1lbSQXjQZiP2F18xjomFbZR-zJ_tr_HB-Iy0)

The problem arises from these codes. What's odd is that in the `DefaultTCPIPHandler` flow, I can only find the 'metadata restore' and cannot locate where the 'metadata store' takes place.

https://github.com/shellhub-io/shellhub/blob/0674b42c6b9372e8fa3745861d113baad034bac9/ssh/server/channels/tcpip.go#L52-L56

https://github.com/shellhub-io/shellhub/blob/0674b42c6b9372e8fa3745861d113baad034bac9/ssh/server/channels/tcpip.go#L60-L65

I discovered that the code for 'metadata store' only executes when connecting through the SSH session flow.

https://github.com/shellhub-io/shellhub/blob/0674b42c6b9372e8fa3745861d113baad034bac9/ssh/server/handler/ssh.go#L170

https://github.com/shellhub-io/shellhub/blob/0674b42c6b9372e8fa3745861d113baad034bac9/ssh/server/handler/ssh.go#L185

Please check if there's anything I might have missed.

Thank you.

### Edition

Community

### Version

v0.13.0-rc.11

--- a/ssh/server/channels/tcpip.go
+++ b/ssh/server/channels/tcpip.go
@@ -6,7 +6,10 @@ import (
 	"strconv"
 
 	gliderssh "github.com/gliderlabs/ssh"
+	"github.com/shellhub-io/shellhub/pkg/httptunnel"
 	"github.com/shellhub-io/shellhub/ssh/pkg/metadata"
+	"github.com/shellhub-io/shellhub/ssh/session"
+	log "github.com/sirupsen/logrus"
 	gossh "golang.org/x/crypto/ssh"
 )
 
@@ -24,68 +27,164 @@ const (
 	DirectTCPIPChannel = "direct-tcpip"
 )
 
-// DefaultTCPIPHandler is the default handler for DirectTCPIPChannel and DynamicTCPIPChannel channels.
-//
-// It will reject the channel if the LocalPortForwardingCallback is not set or returns false.
-// Otherwise, it will dial the agent and proxy the channel.
-func DefaultTCPIPHandler(server *gliderssh.Server, _ *gossh.ServerConn, newChan gossh.NewChannel, ctx gliderssh.Context) {
-	type channelData struct {
-		DestAddr   string
-		DestPort   uint32
-		OriginAddr string
-		OriginPort uint32
+func TunnelDefaultDirectTCPIPHandler(tunnel *httptunnel.Tunnel) func(server *gliderssh.Server, _ *gossh.ServerConn, newChan gossh.NewChannel, ctx gliderssh.Context) {
+	return func(server *gliderssh.Server, conn *gossh.ServerConn, newChan gossh.NewChannel, ctx gliderssh.Context) {
+		target := metadata.RestoreTarget(ctx)
+
+		log.WithFields(log.Fields{
+			"username": target.Username,
+			"sshid":    target.Data,
+		}).Info("handling direct-tcpip channel")
+
+		type channelData struct {
+			DestAddr   string
+			DestPort   uint32
+			OriginAddr string
+			OriginPort uint32
+		}
+
+		data := new(channelData)
+		if err := gossh.Unmarshal(newChan.ExtraData(), data); err != nil {
+			newChan.Reject(gossh.ConnectionFailed, "faild to parse forward data: "+err.Error()) //nolint:errcheck
+			log.WithError(err).WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Error("failed to parse forward data")
+
+			return
+		}
+
+		if server.LocalPortForwardingCallback == nil || !server.LocalPortForwardingCallback(ctx, data.DestAddr, data.DestPort) {
+			newChan.Reject(gossh.Prohibited, "port forwarding is disabled") //nolint:errcheck
+			log.WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Info("port forwarding is disabled")
+
+			return
+		}
+
+		dest := net.JoinHostPort(data.DestAddr, strconv.FormatInt(int64(data.DestPort), 10))
+		config, err := session.NewClientConfiguration(ctx)
+		if err != nil {
+			newChan.Reject(gossh.ConnectionFailed, "error creating client configuration: "+err.Error()) //nolint:errcheck
+			log.WithError(err).WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Error("error creating client configuration")
+
+			return
+		}
+
+		sess, err := session.NewSessionWithoutClient(ctx, tunnel)
+		if err != nil {
+			newChan.Reject(gossh.ConnectionFailed, "failed to create session") //nolint:errcheck
+			log.WithError(err).WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Error("failed to create session")
+
+			return
+		}
+
+		connection, _, err := sess.NewClientConnWithDeadline(config)
+		if err != nil {
+			newChan.Reject(gossh.ConnectionFailed, "failed creating client connection: "+err.Error()) //nolint:errcheck
+			log.WithError(err).WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Error("failed creating agent connection")
+
+			return
+		}
+
+		agent, err := connection.Dial("tcp", dest)
+		if err != nil {
+			newChan.Reject(gossh.ConnectionFailed, "failed dialing the agent to host and port: "+err.Error()) //nolint:errcheck
+			log.WithError(err).WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Error("failed dialing the agent to host and port")
+
+			return
+		}
+
+		channel, reqs, err := newChan.Accept()
+		if err != nil {
+			newChan.Reject(gossh.ConnectionFailed, "failed accepting the channel: "+err.Error()) //nolint:errcheck
+			log.WithError(err).WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Error("failed accepting the channel")
+
+			return
+		}
+
+		go gossh.DiscardRequests(reqs)
+
+		log.WithFields(log.Fields{
+			"username":    target.Username,
+			"sshid":       target.Data,
+			"origin_port": data.OriginAddr,
+			"origin_addr": data.OriginPort,
+			"dest_port":   data.DestPort,
+			"dest_addr":   data.DestAddr,
+		}).Info("piping data between client and agent")
+
+		// TODO: control the running state of these goroutines.
+		go func() {
+			log.WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Debug("copying data from client to agent")
+
+			defer channel.Close()
+			io.Copy(channel, agent) //nolint:errcheck
+		}()
+		go func() {
+			log.WithFields(log.Fields{
+				"username":    target.Username,
+				"sshid":       target.Data,
+				"origin_port": data.OriginAddr,
+				"origin_addr": data.OriginPort,
+				"dest_port":   data.DestPort,
+				"dest_addr":   data.DestAddr,
+			}).Debug("copying data from agent to client")
+
+			defer channel.Close()
+			io.Copy(agent, channel) //nolint:errcheck
+		}()
 	}
-
-	data := channelData{}
-	if err := gossh.Unmarshal(newChan.ExtraData(), &data); err != nil {
-		newChan.Reject(gossh.ConnectionFailed, "error parsing forward data: "+err.Error()) //nolint:errcheck
-
-		return
-	}
-
-	if server.LocalPortForwardingCallback == nil || !server.LocalPortForwardingCallback(ctx, data.DestAddr, data.DestPort) {
-		newChan.Reject(gossh.Prohibited, "port forwarding is disabled") //nolint:errcheck
-
-		return
-	}
-
-	if !metadata.RestoreEstablished(ctx) {
-		newChan.Reject(gossh.Prohibited, "connection between server and agent is not established yet") //nolint:errcheck
-
-		return
-	}
-
-	dest := net.JoinHostPort(data.DestAddr, strconv.FormatInt(int64(data.DestPort), 10))
-
-	agent := metadata.RestoreAgent(ctx)
-	if agent == nil {
-		newChan.Reject(gossh.ConnectionFailed, "error restoring the agent") //nolint:errcheck
-
-		return
-	}
-
-	dialed, err := agent.Dial("tcp", dest)
-	if err != nil {
-		newChan.Reject(gossh.ConnectionFailed, "error dialing the agent to host and port: "+err.Error()) //nolint:errcheck
-
-		return
-	}
-
-	channel, reqs, err := newChan.Accept()
-	if err != nil {
-		newChan.Reject(gossh.ConnectionFailed, "error accepting the channel: "+err.Error()) //nolint:errcheck
-
-		return
-	}
-
-	go gossh.DiscardRequests(reqs)
-
-	go func() {
-		defer channel.Close()
-		io.Copy(channel, dialed) //nolint:errcheck
-	}()
-	go func() {
-		defer channel.Close()
-		io.Copy(dialed, channel) //nolint:errcheck
-	}()
 }

--- a/ssh/server/handler/sftp.go
+++ b/ssh/server/handler/sftp.go
@@ -42,7 +42,7 @@ func SFTPSubsystemHandler(tunnel *httptunnel.Tunnel) gliderssh.SubsystemHandler 
 
 		config, err := session.NewClientConfiguration(ctx)
 		if err != nil {
-			writeError(sess, "Error while creating client configuration", err, err)
+			writeError(sess, "Error while creating client configuration", err, ErrConfiguration)
 
 			return
 		}

--- a/ssh/server/handler/sftp.go
+++ b/ssh/server/handler/sftp.go
@@ -2,8 +2,6 @@ package handler
 
 import (
 	"context"
-	"crypto/x509"
-	"encoding/pem"
 	"fmt"
 
 	gliderssh "github.com/gliderlabs/ssh"
@@ -42,45 +40,11 @@ func SFTPSubsystemHandler(tunnel *httptunnel.Tunnel) gliderssh.SubsystemHandler 
 
 		defer sess.Finish() // nolint:errcheck
 
-		config := &gossh.ClientConfig{ // nolint: exhaustruct
-			User:            sess.Username,
-			HostKeyCallback: gossh.InsecureIgnoreHostKey(), // nolint:gosec
-		}
+		config, err := session.NewClientConfiguration(ctx)
+		if err != nil {
+			writeError(sess, "Error while creating client configuration", err, err)
 
-		switch metadata.RestoreAuthenticationMethod(ctx) {
-		case metadata.PublicKeyAuthenticationMethod:
-			privateKey, err := api.CreatePrivateKey()
-			if err != nil {
-				writeError(sess, "Error while creating private key", err, ErrPrivateKey)
-
-				return
-			}
-
-			block, _ := pem.Decode(privateKey.Data)
-
-			parsed, err := x509.ParsePKCS1PrivateKey(block.Bytes)
-			if err != nil {
-				writeError(sess, "Error while parsing private key", err, ErrPublicKey)
-
-				return
-			}
-
-			signer, err := gossh.NewSignerFromKey(parsed)
-			if err != nil {
-				writeError(sess, "Error while creating signer from private key", err, ErrSigner)
-
-				return
-			}
-
-			config.Auth = []gossh.AuthMethod{
-				gossh.PublicKeys(signer),
-			}
-		case metadata.PasswordAuthenticationMethod:
-			password := metadata.RestorePassword(ctx)
-
-			config.Auth = []gossh.AuthMethod{
-				gossh.Password(password),
-			}
+			return
 		}
 
 		if err = connectSFTP(ctx, client, sess, api, config); err != nil {

--- a/ssh/server/handler/ssh.go
+++ b/ssh/server/handler/ssh.go
@@ -3,8 +3,6 @@ package handler
 import (
 	"bytes"
 	"context"
-	"crypto/x509"
-	"encoding/pem"
 	"errors"
 	"fmt"
 	"io"
@@ -81,6 +79,13 @@ func SSHHandler(tunnel *httptunnel.Tunnel) gliderssh.Handler {
 		}
 		defer sess.Finish() // nolint: errcheck
 
+		opts, err := envs.ParseWithPrefix[ConfigOptions]("")
+		if err != nil {
+			writeError(sess, "Error while parsing envs", err, ErrEnvs)
+
+			return
+		}
+
 		if wh := webhook.NewClient(); wh != nil {
 			res, err := wh.Connect(sess.Lookup)
 			if errors.Is(err, webhook.ErrForbidden) {
@@ -95,58 +100,16 @@ func SSHHandler(tunnel *httptunnel.Tunnel) gliderssh.Handler {
 
 			time.Sleep(time.Duration(res.Timeout) * time.Second)
 		}
+		ctx := client.Context()
 
-		opts, err := envs.ParseWithPrefix[ConfigOptions]("")
+		config, err := session.NewClientConfiguration(ctx)
 		if err != nil {
-			writeError(sess, "Error while parsing envs", err, ErrEnvs)
+			writeError(sess, "Error while creating client configuration", err, err)
 
 			return
 		}
 
-		config := &gossh.ClientConfig{ // nolint: exhaustruct
-			User:            sess.Username,
-			HostKeyCallback: gossh.InsecureIgnoreHostKey(), // nolint: gosec
-		}
-
-		ctx := client.Context()
 		api := metadata.RestoreAPI(ctx)
-
-		switch metadata.RestoreAuthenticationMethod(ctx) {
-		case metadata.PublicKeyAuthenticationMethod:
-			privateKey, err := api.CreatePrivateKey()
-			if err != nil {
-				writeError(sess, "Error while creating private key", err, ErrPrivateKey)
-
-				return
-			}
-
-			block, _ := pem.Decode(privateKey.Data)
-
-			parsed, err := x509.ParsePKCS1PrivateKey(block.Bytes)
-			if err != nil {
-				writeError(sess, "Error while parsing private key", err, ErrPublicKey)
-
-				return
-			}
-
-			signer, err := gossh.NewSignerFromKey(parsed)
-			if err != nil {
-				writeError(sess, "Error while creating signer from private key", err, ErrSigner)
-
-				return
-			}
-
-			config.Auth = []gossh.AuthMethod{
-				gossh.PublicKeys(signer),
-			}
-		case metadata.PasswordAuthenticationMethod:
-			password := metadata.RestorePassword(ctx)
-
-			config.Auth = []gossh.AuthMethod{
-				gossh.Password(password),
-			}
-		}
-
 		err = connectSSH(ctx, client, sess, config, api, *opts)
 		if err != nil {
 			writeError(sess, "Error during SSH connection", err, err)
@@ -177,7 +140,6 @@ func connectSSH(ctx context.Context, client gliderssh.Session, sess *session.Ses
 
 		return ErrSession
 	}
-
 	defer agent.Close()
 
 	go session.HandleRequests(ctx, reqs, api, ctx.Done())

--- a/ssh/server/handler/ssh.go
+++ b/ssh/server/handler/ssh.go
@@ -53,6 +53,7 @@ var (
 	ErrTarget             = fmt.Errorf("failed to get client target")
 	ErrAuthentication     = fmt.Errorf("failed to authenticate to device")
 	ErrEnvs               = fmt.Errorf("failed to parse server envs")
+	ErrConfiguration      = fmt.Errorf("failed to create communication configuration")
 )
 
 type ConfigOptions struct {
@@ -104,7 +105,7 @@ func SSHHandler(tunnel *httptunnel.Tunnel) gliderssh.Handler {
 
 		config, err := session.NewClientConfiguration(ctx)
 		if err != nil {
-			writeError(sess, "Error while creating client configuration", err, err)
+			writeError(sess, "Error while creating client configuration", err, ErrConfiguration)
 
 			return
 		}

--- a/ssh/server/server.go
+++ b/ssh/server/server.go
@@ -40,6 +40,13 @@ func NewServer(opts *Options, tunnel *httptunnel.Tunnel) *Server {
 		SessionRequestCallback: func(client gliderssh.Session, request string) bool {
 			metadata.StoreRequest(client.Context(), request)
 
+			target := metadata.RestoreTarget(client.Context())
+			log.WithFields(log.Fields{
+				"username": target.Username,
+				"sshid":    target.Data,
+				"request":  request,
+			}).Info("Session request")
+
 			return true
 		},
 		Handler: handler.SSHHandler(tunnel),
@@ -54,7 +61,7 @@ func NewServer(opts *Options, tunnel *httptunnel.Tunnel) *Server {
 		},
 		ChannelHandlers: map[string]gliderssh.ChannelHandler{
 			"session":                   gliderssh.DefaultSessionHandler,
-			channels.DirectTCPIPChannel: channels.DefaultTCPIPHandler,
+			channels.DirectTCPIPChannel: channels.TunnelDefaultDirectTCPIPHandler(tunnel),
 		},
 	}
 

--- a/ssh/session/errors.go
+++ b/ssh/session/errors.go
@@ -1,0 +1,14 @@
+package session
+
+import "fmt"
+
+// Errors returned by the NewSession to the client.
+var (
+	ErrBillingBlock       = fmt.Errorf("Connection to this device is not available as your current namespace doesn't qualify for the free plan. To gain access, you'll need to contact the namespace owner to initiate an upgrade.\n\nFor a detailed estimate of costs based on your use-cases with ShellHub Cloud, visit our pricing page at https://www.shellhub.io/pricing. If you wish to upgrade immediately, navigate to https://cloud.shellhub.io/settings/billing. Your cooperation is appreciated.") //nolint:all
+	ErrFirewallBlock      = fmt.Errorf("you cannot connect to this device because a firewall rule block your connection")
+	ErrFirewallConnection = fmt.Errorf("failed to communicate to the firewall")
+	ErrFirewallUnknown    = fmt.Errorf("failed to evaluate the firewall rule")
+	ErrHost               = fmt.Errorf("failed to get the device address")
+	ErrFindDevice         = fmt.Errorf("failed to find the device")
+	ErrDial               = fmt.Errorf("failed to connect to device agent, please check the device connection")
+)

--- a/ssh/session/session.go
+++ b/ssh/session/session.go
@@ -158,6 +158,57 @@ func NewSession(client gliderssh.Session, tunnel *httptunnel.Tunnel) (*Session, 
 	return session, nil
 }
 
+// NewSessionWithoutClient creates a new session to connect the agent, validating data, instance and payment.
+//
+// This function is used to create a new session when the client is not available, what is true when the SSH client
+// indicate that the request type is `none` or in the case of a port forwarding
+func NewSessionWithoutClient(ctx gliderssh.Context, tunnel *httptunnel.Tunnel) (*Session, error) {
+	uid := ctx.Value(gliderssh.ContextKeySessionID).(string) //nolint:forcetypeassert
+
+	device := metadata.RestoreDevice(ctx)
+	target := metadata.RestoreTarget(ctx)
+	lookup := metadata.RestoreLookup(ctx)
+
+	lookup["username"] = target.Username
+	// lookup["ip_address"] = hos.Host
+
+	session := new(Session)
+	if ok, err := session.checkFirewall(ctx); err != nil || !ok {
+		log.WithError(err).
+			WithFields(log.Fields{"session": uid, "sshid": target.Username}).
+			Error("Error when trying to evaluate firewall rules")
+
+		return nil, err
+	}
+
+	if ok, err := session.checkBilling(ctx, device.UID); err != nil || !ok {
+		log.WithError(err).
+			WithFields(log.Fields{"session": uid, "sshid": target.Username}).
+			Error("Error when trying to evaluate billing")
+
+		return nil, err
+	}
+
+	dialed, err := session.dial(ctx, tunnel, device.UID, uid)
+	if err != nil {
+		log.WithError(err).
+			WithFields(log.Fields{"session": uid, "sshid": target.Username}).
+			Error("Error when trying to dial")
+
+		return nil, ErrDial
+	}
+
+	session.Client = nil
+	session.UID = uid
+	session.Username = target.Username
+	// session.IPAddress = hos.Host
+	session.Device = device.UID
+	session.Lookup = lookup
+	session.Dialed = dialed
+
+	return session, nil
+}
+
 func (s *Session) GetType() string {
 	return s.Type
 }

--- a/ssh/session/session.go
+++ b/ssh/session/session.go
@@ -1,6 +1,8 @@
 package session
 
 import (
+	"crypto/x509"
+	"encoding/pem"
 	"errors"
 	"fmt"
 	"net"
@@ -35,8 +37,64 @@ type Session struct {
 	Dialed        net.Conn
 }
 
+// checkFirewall evaluates if there are firewall rules that block the connection.
+func (s *Session) checkFirewall(ctx gliderssh.Context) (bool, error) {
+	api := metadata.RestoreAPI(ctx)
+	lookup := metadata.RestoreLookup(ctx)
+
+	if envs.IsCloud() || envs.IsEnterprise() {
+		if err := api.FirewallEvaluate(lookup); err != nil {
+			switch {
+			case errors.Is(err, internalclient.ErrFirewallConnection):
+				return false, errors.Join(ErrFirewallConnection, err)
+			case errors.Is(err, internalclient.ErrFirewallBlock):
+				return false, errors.Join(ErrFirewallBlock, err)
+			default:
+				return false, errors.Join(ErrFirewallUnknown, err)
+			}
+		}
+	}
+
+	return true, nil
+}
+
+// checkBilling evaluates if the device's namespace has pending payment questions.
+func (s *Session) checkBilling(ctx gliderssh.Context, device string) (bool, error) {
+	api := metadata.RestoreAPI(ctx)
+
+	if envs.IsCloud() && envs.HasBilling() {
+		device, err := api.GetDevice(device)
+		if err != nil {
+			return false, errors.Join(ErrFindDevice, err)
+		}
+
+		if evaluatation, status, _ := api.BillingEvaluate(device.TenantID); status != 402 && !evaluatation.CanConnect {
+			return false, errors.Join(ErrBillingBlock, err)
+		}
+	}
+
+	return true, nil
+}
+
+// dial dials the a connection between SSH server and the device agent.
+func (s *Session) dial(ctx gliderssh.Context, tunnel *httptunnel.Tunnel, device string, session string) (net.Conn, error) {
+	dialed, err := tunnel.Dial(ctx, device)
+	if err != nil {
+		return nil, errors.Join(ErrDial, err)
+	}
+
+	req, _ := http.NewRequest(http.MethodGet, fmt.Sprintf("/ssh/%s", session), nil)
+	if err = req.Write(dialed); err != nil {
+		return nil, err
+	}
+
+	return dialed, nil
+}
+
 // NewSession creates a new Client from a client to agent, validating data, instance and payment.
 func NewSession(client gliderssh.Session, tunnel *httptunnel.Tunnel) (*Session, error) {
+	ctx := client.Context()
+
 	hos, err := host.NewHost(client.RemoteAddr().String())
 	if err != nil {
 		return nil, ErrHost
@@ -49,54 +107,33 @@ func NewSession(client gliderssh.Session, tunnel *httptunnel.Tunnel) (*Session, 
 		}
 	}
 
-	clientCtx := client.Context()
+	uid := ctx.Value(gliderssh.ContextKeySessionID).(string) //nolint:forcetypeassert
 
-	uid := clientCtx.Value(gliderssh.ContextKeySessionID).(string) //nolint:forcetypeassert
-	device := metadata.RestoreDevice(clientCtx)
-	tag := metadata.RestoreTarget(clientCtx)
-	api := metadata.RestoreAPI(clientCtx)
-	lookup := metadata.RestoreLookup(clientCtx)
+	device := metadata.RestoreDevice(ctx)
+	target := metadata.RestoreTarget(ctx)
+	lookup := metadata.RestoreLookup(ctx)
 
-	lookup["username"] = tag.Username
+	lookup["username"] = target.Username
 	lookup["ip_address"] = hos.Host
 
-	if envs.IsCloud() || envs.IsEnterprise() {
-		if err := api.FirewallEvaluate(lookup); err != nil {
-			log.WithError(err).
-				WithFields(log.Fields{"session": uid, "sshid": client.User()}).
-				Error("Error when trying to evaluate firewall rules")
+	session := new(Session)
+	if ok, err := session.checkFirewall(ctx); err != nil || !ok {
+		log.WithError(err).
+			WithFields(log.Fields{"session": uid, "sshid": client.User()}).
+			Error("Error when trying to evaluate firewall rules")
 
-			switch {
-			case errors.Is(err, internalclient.ErrFirewallConnection):
-				return nil, ErrFirewallConnection
-			case errors.Is(err, internalclient.ErrFirewallBlock):
-				return nil, ErrFirewallBlock
-			default:
-				return nil, ErrFirewallUnknown
-			}
-		}
+		return nil, err
 	}
 
-	if envs.IsCloud() && envs.HasBilling() {
-		device, err := api.GetDevice(device.UID)
-		if err != nil {
-			log.WithError(err).
-				WithFields(log.Fields{"session": uid, "sshid": client.User()}).
-				Error("Error when trying to get device")
+	if ok, err := session.checkBilling(ctx, device.UID); err != nil || !ok {
+		log.WithError(err).
+			WithFields(log.Fields{"session": uid, "sshid": client.User()}).
+			Error("Error when trying to evaluate billing")
 
-			return nil, ErrFindDevice
-		}
-
-		if evaluatation, status, _ := api.BillingEvaluate(device.TenantID); status != 402 && !evaluatation.CanConnect {
-			log.WithError(err).
-				WithFields(log.Fields{"session": uid, "sshid": client.User()}).
-				Error("Error when trying to evaluate billing")
-
-			return nil, ErrBillingBlock
-		}
+		return nil, err
 	}
 
-	dialed, err := tunnel.Dial(client.Context(), device.UID)
+	dialed, err := session.dial(ctx, tunnel, device.UID, uid)
 	if err != nil {
 		log.WithError(err).
 			WithFields(log.Fields{"session": uid, "sshid": client.User()}).
@@ -105,24 +142,13 @@ func NewSession(client gliderssh.Session, tunnel *httptunnel.Tunnel) (*Session, 
 		return nil, ErrDial
 	}
 
-	req, _ := http.NewRequest(http.MethodGet, fmt.Sprintf("/ssh/%s", uid), nil)
-	if err = req.Write(dialed); err != nil {
-		log.WithError(err).
-			WithFields(log.Fields{"session": uid, "sshid": client.User()}).
-			Error("Error when trying to write the request")
-
-		return nil, err
-	}
-
-	session := &Session{ //nolint:exhaustruct
-		Client:    client,
-		UID:       uid,
-		Username:  tag.Username,
-		IPAddress: hos.Host,
-		Device:    device.UID,
-		Lookup:    lookup,
-		Dialed:    dialed,
-	}
+	session.Client = client
+	session.UID = uid
+	session.Username = target.Username
+	session.IPAddress = hos.Host
+	session.Device = device.UID
+	session.Lookup = lookup
+	session.Dialed = dialed
 
 	session.setPty()
 	session.setType()
@@ -153,7 +179,7 @@ func (s *Session) NewClientConnWithDeadline(config *gossh.ClientConfig) (*gossh.
 	cli, chans, reqs, err := gossh.NewClientConn(s.Dialed, Addr, config)
 	if err != nil {
 		log.WithError(err).
-			WithFields(log.Fields{"session": s.UID, "sshid": s.Client.User()}).
+			WithFields(log.Fields{"session": s.UID}).
 			Error("Error when trying to create the client's connection")
 
 		return nil, nil, err

--- a/ssh/session/session.go
+++ b/ssh/session/session.go
@@ -1,12 +1,10 @@
 package session
 
 import (
-	"context"
 	"errors"
 	"fmt"
 	"net"
 	"net/http"
-	"strings"
 	"time"
 
 	gliderssh "github.com/gliderlabs/ssh"
@@ -19,17 +17,6 @@ import (
 	"github.com/shellhub-io/shellhub/ssh/pkg/metadata"
 	log "github.com/sirupsen/logrus"
 	gossh "golang.org/x/crypto/ssh"
-)
-
-// Errors returned by the NewSession to the client.
-var (
-	ErrBillingBlock       = fmt.Errorf("Connection to this device is not available as your current namespace doesn't qualify for the free plan. To gain access, you'll need to contact the namespace owner to initiate an upgrade.\n\nFor a detailed estimate of costs based on your use-cases with ShellHub Cloud, visit our pricing page at https://www.shellhub.io/pricing. If you wish to upgrade immediately, navigate to https://cloud.shellhub.io/settings/billing. Your cooperation is appreciated.") //nolint:all
-	ErrFirewallBlock      = fmt.Errorf("you cannot connect to this device because a firewall rule block your connection")
-	ErrFirewallConnection = fmt.Errorf("failed to communicate to the firewall")
-	ErrFirewallUnknown    = fmt.Errorf("failed to evaluate the firewall rule")
-	ErrHost               = fmt.Errorf("failed to get the device address")
-	ErrFindDevice         = fmt.Errorf("failed to find the device")
-	ErrDial               = fmt.Errorf("failed to connect to device agent, please check the device connection")
 )
 
 type Session struct {
@@ -46,66 +33,6 @@ type Session struct {
 	Lookup        map[string]string
 	Pty           bool
 	Dialed        net.Conn
-}
-
-const (
-	Web     = "web"     // web terminal.
-	Term    = "term"    // interactive session
-	Exec    = "exec"    // command execution
-	HereDoc = "heredoc" // heredoc pty.
-	SCP     = "scp"     // scp.
-	SFTP    = "sftp"    // sftp subsystem.
-	Unk     = "unknown" // unknown.
-)
-
-// setPty sets the connection's pty.
-func (s *Session) setPty() {
-	pty, _, isPty := s.Client.Pty()
-	if isPty {
-		s.Term = pty.Term
-	}
-
-	s.Pty = isPty
-}
-
-// setType sets the connection`s type to session.
-//
-// Connection types possible are: Web, SFTP, SCP, Exec, HereDoc, Term, Unk (unknown)
-func (s *Session) setType() {
-	ctx := s.Client.Context()
-
-	env := loadEnv(s.Client.Environ())
-	if value, ok := env["WS"]; ok && value == "true" {
-		env["WS"] = "false"
-		s.Type = Web
-
-		return
-	}
-
-	if s.Client.Subsystem() == SFTP {
-		s.Type = SFTP
-
-		return
-	}
-
-	var cmd string
-	commands := s.Client.Command()
-	if len(commands) != 0 {
-		cmd = commands[0]
-	}
-
-	switch {
-	case !s.Pty && strings.HasPrefix(cmd, SCP):
-		s.Type = SCP
-	case !s.Pty && metadata.RestoreRequest(ctx) == "shell":
-		s.Type = HereDoc
-	case cmd != "":
-		s.Type = Exec
-	case s.Pty:
-		s.Type = Term
-	default:
-		s.Type = Unk
-	}
 }
 
 // NewSession creates a new Client from a client to agent, validating data, instance and payment.
@@ -282,51 +209,4 @@ func (s *Session) Finish() error {
 	}
 
 	return nil
-}
-
-func loadEnv(env []string) map[string]string {
-	m := make(map[string]string, cap(env))
-
-	for _, s := range env {
-		sp := strings.Split(s, "=")
-		if len(sp) == 2 {
-			k := sp[0]
-			v := sp[1]
-			m[k] = v
-		}
-	}
-
-	return m
-}
-
-func HandleRequests(ctx context.Context, reqs <-chan *gossh.Request, c internalclient.Client, done <-chan struct{}) {
-	for {
-		select {
-		case req := <-reqs:
-			if req == nil {
-				break
-			}
-
-			switch req.Type {
-			case "keepalive":
-				if id, ok := ctx.Value(gliderssh.ContextKeySessionID).(string); ok {
-					if errs := c.KeepAliveSession(id); len(errs) > 0 {
-						log.Error(errs[0])
-					}
-				}
-
-				if err := req.Reply(false, nil); err != nil {
-					log.Error(err)
-				}
-			default:
-				if req.WantReply {
-					if err := req.Reply(false, nil); err != nil {
-						log.Error(err)
-					}
-				}
-			}
-		case <-done:
-			return
-		}
-	}
 }

--- a/ssh/session/session.go
+++ b/ssh/session/session.go
@@ -236,3 +236,54 @@ func (s *Session) Finish() error {
 
 	return nil
 }
+
+// NewClientConfiguration creates a [gossh.ClientConfig] with the default configuration required by ShellHub
+// to connect to the device agent that are inside the [gliderssh.Context].
+func NewClientConfiguration(ctx gliderssh.Context) (*gossh.ClientConfig, error) {
+	target := metadata.RestoreTarget(ctx)
+	if target == nil {
+		return nil, errors.New("failed to get the target from context")
+	}
+
+	config := &gossh.ClientConfig{
+		User:            target.Username,
+		HostKeyCallback: gossh.InsecureIgnoreHostKey(), // nolint: gosec
+	}
+
+	api := metadata.RestoreAPI(ctx)
+	if api == nil {
+		return nil, errors.New("failed to get the API from context")
+	}
+
+	switch metadata.RestoreAuthenticationMethod(ctx) {
+	case metadata.PublicKeyAuthenticationMethod:
+		privateKey, err := api.CreatePrivateKey()
+		if err != nil {
+			return nil, err
+		}
+
+		block, _ := pem.Decode(privateKey.Data)
+
+		parsed, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		signer, err := gossh.NewSignerFromKey(parsed)
+		if err != nil {
+			return nil, err
+		}
+
+		config.Auth = []gossh.AuthMethod{
+			gossh.PublicKeys(signer),
+		}
+	case metadata.PasswordAuthenticationMethod:
+		password := metadata.RestorePassword(ctx)
+
+		config.Auth = []gossh.AuthMethod{
+			gossh.Password(password),
+		}
+	}
+
+	return config, nil
+}

--- a/ssh/session/types.go
+++ b/ssh/session/types.go
@@ -1,0 +1,67 @@
+package session
+
+import (
+	"strings"
+
+	"github.com/shellhub-io/shellhub/ssh/pkg/metadata"
+)
+
+const (
+	Web     = "web"     // web terminal.
+	Term    = "term"    // interactive session
+	Exec    = "exec"    // command execution
+	HereDoc = "heredoc" // heredoc pty.
+	SCP     = "scp"     // scp.
+	SFTP    = "sftp"    // sftp subsystem.
+	Unk     = "unknown" // unknown.
+)
+
+// setPty sets the connection's pty.
+func (s *Session) setPty() {
+	pty, _, isPty := s.Client.Pty()
+	if isPty {
+		s.Term = pty.Term
+	}
+
+	s.Pty = isPty
+}
+
+// setType sets the connection`s type to session.
+//
+// Connection types possible are: Web, SFTP, SCP, Exec, HereDoc, Term, Unk (unknown)
+func (s *Session) setType() {
+	ctx := s.Client.Context()
+
+	env := loadEnv(s.Client.Environ())
+	if value, ok := env["WS"]; ok && value == "true" {
+		env["WS"] = "false"
+		s.Type = Web
+
+		return
+	}
+
+	if s.Client.Subsystem() == SFTP {
+		s.Type = SFTP
+
+		return
+	}
+
+	var cmd string
+	commands := s.Client.Command()
+	if len(commands) != 0 {
+		cmd = commands[0]
+	}
+
+	switch {
+	case !s.Pty && strings.HasPrefix(cmd, SCP):
+		s.Type = SCP
+	case !s.Pty && metadata.RestoreRequest(ctx) == "shell":
+		s.Type = HereDoc
+	case cmd != "":
+		s.Type = Exec
+	case s.Pty:
+		s.Type = Term
+	default:
+		s.Type = Unk
+	}
+}

--- a/ssh/session/utils.go
+++ b/ssh/session/utils.go
@@ -1,0 +1,59 @@
+package session
+
+import (
+	"context"
+	"strings"
+
+	gliderssh "github.com/gliderlabs/ssh"
+	"github.com/shellhub-io/shellhub/pkg/api/internalclient"
+	log "github.com/sirupsen/logrus"
+	gossh "golang.org/x/crypto/ssh"
+)
+
+// TODO: Evaluate if we can use a dedicated package for this.
+func loadEnv(env []string) map[string]string {
+	m := make(map[string]string, cap(env))
+
+	for _, s := range env {
+		sp := strings.Split(s, "=")
+		if len(sp) == 2 {
+			k := sp[0]
+			v := sp[1]
+			m[k] = v
+		}
+	}
+
+	return m
+}
+
+func HandleRequests(ctx context.Context, reqs <-chan *gossh.Request, c internalclient.Client, done <-chan struct{}) {
+	for {
+		select {
+		case req := <-reqs:
+			if req == nil {
+				break
+			}
+
+			switch req.Type {
+			case "keepalive":
+				if id, ok := ctx.Value(gliderssh.ContextKeySessionID).(string); ok {
+					if errs := c.KeepAliveSession(id); len(errs) > 0 {
+						log.Error(errs[0])
+					}
+				}
+
+				if err := req.Reply(false, nil); err != nil {
+					log.Error(err)
+				}
+			default:
+				if req.WantReply {
+					if err := req.Reply(false, nil); err != nil {
+						log.Error(err)
+					}
+				}
+			}
+		case <-done:
+			return
+		}
+	}
+}

